### PR TITLE
Add FreeBSD/powerpc64le support

### PR DESCRIPTION
**Pull request by pkubaj**
### Changelog category (leave one):
- Buld/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add support for FreeBSD/powerpc64le.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Ok, but we also need to put the files here: https://github.com/ClickHouse/sysroot

**Review by alexey-milovidov — APPROVED** [review]
The changes LGTM.

**Comment by pkubaj** [comment]
You can get it at https://pkg.freebsd.org/FreeBSD:14:amd64/latest/All/powerpc64le-freebsd-sysroot-a2021.11.09.pkg. This is a FreeBSD package, but it's just a standard tar.zstd renamed to pkg. You can extract it with tar.

**Comment by pkubaj** [comment]
Oh, I guess that won't do. This is a sysroot for FreeBSD 13.0, on which LLVM doesn't have SSE translation (-DNO_WARN_X86_INTRINSICS stuff). To build clickhouse myself, I used 14.0 snapshot. So I guess at this time we should ditch adding a sysroot, since it won't work anyway.

**Comment by alexey-milovidov** [comment]
We don't use any "tools" from the toolchain, only the libc headers, and blobs, and we compile everything with the host clang/lld/LLVM tools (which already support cross-compilation).

**Comment by pkubaj** [comment]
OK, then it should be fine.

**Comment by alexey-milovidov** [comment]
Could you please add the files to https://github.com/ClickHouse/sysroot ?

**Comment by pkubaj** [comment]
One more thing: CI currently won't pass on FreeBSD/powerpc64le. The reason is that more patches are necessary (FreeBSD-specific), but those are only for the sources in contrib (e.g. abseil-cpp, boringssl and zlib-ng). Since you provide your own forks of those, are you willing to accept patches to those?

**Comment by alexey-milovidov** [comment]
@pkubaj Yes, we will merge the patches to contrib libraries!

**Comment by pkubaj** [comment]
I noticed that abseil-cpp is from the upstream repo, so unfortunately the patch against it can't be committed to clickhouse's repo. The patch is pretty simple and only affects the headers:
```
--- absl/base/internal/unscaledcycleclock.cc.orig       2020-08-09 20:09:49 UTC
+++ absl/base/internal/unscaledcycleclock.cc
@@ -24,8 +24,10 @@
 #ifdef __GLIBC__
 #include <sys/platform/ppc.h>
 #elif defined(__FreeBSD__)
-#include <sys/sysctl.h>
+#include "absl/base/call_once.h"
 #include <sys/types.h>
+#include <sys/sysctl.h>
+#include <threads.h>
 #endif
 #endif
```

As such, would it be possible for you to commit just the clickhouse-related patches that I added in this PR?

**Comment by alexey-milovidov** [comment]
Please submit the patch here: https://github.com/ClickHouse/abseil-cpp
And we will switch the submodule to use this fork.

**Comment by pkubaj** [comment]
Thank you, I have submitted all the necessary patches!

**Comment by alexey-milovidov** [comment]
Ok. Now need to submit the files to https://github.com/ClickHouse/sysroot

**Comment by alexey-milovidov** [comment]
@pkubaj Let's add the required files to https://github.com/ClickHouse/sysroot

**Comment by alexey-milovidov** [comment]
@pkubaj, only a small step remains...

**Comment by pkubaj** [comment]
Can you point out which headers can be removed? I don't want to
accidentaly remove something necessary.

**Comment by alexey-milovidov** [comment]
I will try to make it working...

**Comment by alexey-milovidov** [comment]
@pkubaj it does not build:

```
mkdir build_freebsd_powerpc64le
cd build_freebsd_powerpc64le
CC=clang CXX=clang++ cmake -D CMAKE_TOOLCHAIN_FILE=cmake/freebsd/toolchain-ppc64le.cmake -D ENABLE_LDAP=0 ..
ninja clickhouse
```

```
[91/5580] Building C object contrib/boringssl-cmake/CMakeFiles/_crypto.dir/__/boringssl/crypto/cpu-ppc64le.c.o
FAILED: contrib/boringssl-cmake/CMakeFiles/_crypto.dir/__/boringssl/crypto/cpu-ppc64le.c.o 
/usr/bin/ccache /usr/local/bin/clang --target=powerpc64le-unknown-freebsd13 --sysroot=/home/milovidov/work/ClickHouse/cmake/freebsd/../../contrib/sysroot/freebsd-powerpc64le -DBORINGSSL_IMPLEMENTATION -DHAS_RESERVED_IDENTIFIER -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -isystem /home/milovidov/work/ClickHouse/contrib/boringssl/include -isystem /home/milovidov/work/ClickHouse/contrib/llvm-project/libcxxabi/include -isystem /home/milovidov/work/ClickHouse/contrib/libunwind/include -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion  -gdwarf-aranges -pipe -maltivec -mcpu=power8 -D__SSE2__=1 -DNO_WARN_X86_INTRINSICS -fasynchronous-unwind-tables -ffile-prefix-map=/home/milovidov/work/ClickHouse=. -falign-functions=32  -fdiagnostics-absolute-paths -fexperimental-new-pass-manager -w -fno-common -std=c11 -D_XOPEN_SOURCE=700 -O2 -g -DNDEBUG -O3 -g -gdwarf-4    -D OS_FREEBSD -Wno-gnu-anonymous-struct -std=gnu11 -MD -MT contrib/boringssl-cmake/CMakeFiles/_crypto.dir/__/boringssl/crypto/cpu-ppc64le.c.o -MF contrib/boringssl-cmake/CMakeFiles/_crypto.dir/__/boringssl/crypto/cpu-ppc64le.c.o.d -o contrib/boringssl-cmake/CMakeFiles/_crypto.dir/__/boringssl/crypto/cpu-ppc64le.c.o -c /home/milovidov/work/ClickHouse/contrib/boringssl/crypto/cpu-ppc64le.c
/home/milovidov/work/ClickHouse/contrib/boringssl/crypto/cpu-ppc64le.c:31:28: error: call to undeclared function 'getauxval'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  OPENSSL_ppc64le_hwcap2 = getauxval(AT_HWCAP2);
                           ^
1 error generated.
```

File: contrib/jemalloc-cmake/include_freebsd_ppc64le/jemalloc/internal/jemalloc_internal_defs.h.in [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+/* pthread support */
+#define JEMALLOC_HAVE_PTHREAD 
+
+/* dlsym() support */
+#define JEMALLOC_HAVE_DLSYM 
+
+/* Adaptive mutex support in pthreads. */
+#define JEMALLOC_HAVE_PTHREAD_MUTEX_ADAPTIVE_NP 
+
+/* GNU specific sched_getcpu support */
```
Comment (alexey-milovidov) [inline contrib/jemalloc-cmake/include_freebsd_ppc64le/jemalloc/internal/jemalloc_internal_defs.h.in:389]: This is definitely wrong.
How did you generate these files?

**Comment by alexey-milovidov** [comment]
This is not a new feature (edited the changelog entry).

**Comment by alexey-milovidov** [comment]
Did it ever build correctly?

**Comment by pkubaj** [comment]
The older version I tested on did build correctly. It probably didn't have BoringSSL, but since BoringSSL is also used in other projects, I already know the correct fix. I will send a PR.

**Comment by pkubaj** [comment]
https://github.com/ClickHouse/boringssl/pull/16
Now, this is Google we talk about and that patch has 0% chance of being upstreamed.

**Comment by alexey-milovidov** [comment]
Thank you! Merged.

> Now, this is Google we talk about and that patch has 0% chance of being upstreamed.

No problem, it's ok to have patches.

File: contrib/jemalloc-cmake/include_freebsd_ppc64le/jemalloc/internal/jemalloc_internal_defs.h.in (hunk 1 above)
Comment (pkubaj) [reply contrib/jemalloc-cmake/include_freebsd_ppc64le/jemalloc/internal/jemalloc_internal_defs.h.in:389]: I downloaded the latest release of jemalloc and ran the configure script. I just did this again to confirm and there's definitely `#define JEMALLOC_HAVE_PTHREAD_MUTEX_ADAPTIVE_NP`.

Comment (alexey-milovidov) [reply contrib/jemalloc-cmake/include_freebsd_ppc64le/jemalloc/internal/jemalloc_internal_defs.h.in:389]: But it's a non-portable (`NP`) GNU extension for POSIX. It should not exist in FreeBSD.

Comment (alexey-milovidov) [reply contrib/jemalloc-cmake/include_freebsd_ppc64le/jemalloc/internal/jemalloc_internal_defs.h.in:389]: Hm, you are right - it exists on FreeBSD:

```
milovidov@milovidov-desktop:~/work/ClickHouse/contrib/sysroot/freebsd-x86_64$ rg PTHREAD_MUTEX_ADAPTIVE_NP
usr/include/pthread.h
132:    PTHREAD_MUTEX_ADAPTIVE_NP       = 4,    /* Adaptive mutex, spins briefly before blocking on lock */
```

**Pull request merged**